File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.scala
object Solution {
  def strStr(haystack: String, needle: String): Int = {
    val n = haystack.length
    val m = needle.length
    if (m == 0) return 0
    val lps = Array.fill(m)(0)
    var len = 0
    var i = 1
    while (i < m) {
      if (needle(i) == needle(len)) {
        len += 1
        lps(i) = len
        i += 1
      } else if (len > 0) {
        len = lps(len - 1)
      } else {
        lps(i) = 0
        i += 1
      }
    }
    i = 0; var j = 0
    while (i < n) {
      if (haystack(i) == needle(j)) {
        i += 1; j += 1
        if (j == m) return i - m
      } else if (j > 0) {
        j = lps(j - 1)
      } else {
        i += 1
      }
    }
    -1
  }
}

File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.cs
public class Solution {
    public int StrStr(string haystack, string needle) {
        int n = haystack.Length, m = needle.Length;
        if (m == 0) return 0;
        int[] lps = new int[m];
        for (int i = 1, len = 0; i < m; ) {
            if (needle[i] == needle[len]) {
                lps[i++] = ++len;
            } else if (len > 0) {
                len = lps[len - 1];
            } else {
                lps[i++] = 0;
            }
        }
        for (int i = 0, j = 0; i < n; ) {
            if (haystack[i] == needle[j]) {
                i++; j++;
                if (j == m) return i - m;
            } else if (j > 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
        return -1;
    }
}